### PR TITLE
Updated API entities to match v1.0

### PR DESCRIPTION
**Pull request by aasmune**
This PR implements the comments from #98. Could still be defines/constants which may not be used.

### CODE REVIEW

File: DESIGN.md [hunk 1]
```diff
@@ -311,11 +311,11 @@ typedef struct
     /**
      * These fields identify the transfer for the application logic.
      */
-    uint16_t data_type_id;                  ///< 0 to 255 for services, 0 to 65535 for messages
+    uint16_t port_id;                       ///< 0 to 511 for services, 0 to 32767 for messages
     uint8_t transfer_type;                  ///< See @ref CanardTransferType
     uint8_t transfer_id;                    ///< 0 to 31
     uint8_t priority;                       ///< 0 to 31
-    uint8_t source_node_id;                 ///< 1 to 127, or 0 if the source is anonymous
+    uint8_t source_node_id;                 ///< 1 to 127, or 255 if the source is anonymous
```
Comment (pavel-kirienko) [inline DESIGN.md]:
```suggestion
    uint8_t source_node_id;                 ///< 0 to 127, or 255 if the source is anonymous
```

File: canard.h [hunk 2]
```diff
@@ -97,15 +97,15 @@ extern "C" {
...
-
 /// Transfer priority definitions
-#define CANARD_TRANSFER_PRIORITY_HIGHEST            0
-#define CANARD_TRANSFER_PRIORITY_HIGH               8
-#define CANARD_TRANSFER_PRIORITY_MEDIUM             16
-#define CANARD_TRANSFER_PRIORITY_LOW                24
-#define CANARD_TRANSFER_PRIORITY_LOWEST             31
+#define CANARD_TRANSFER_PRIORITY_EXCEPTIONAL    0
+#define CANARD_TRANSFER_PRIORITY_IMMEDIATE      1
+#define CANARD_TRANSFER_FAST                    2
```
Comment (pavel-kirienko) [inline canard.h]:
```suggestion
#define CANARD_TRANSFER_PRIORITY_FAST           2
```

File: canard.h [hunk 3]
```diff
@@ -42,8 +42,8 @@ extern "C" {
 #endif
 
 /// Libcanard version. API will be backwards compatible within the same major version.
-#define CANARD_VERSION_MAJOR                        0
-#define CANARD_VERSION_MINOR                        2
+#define CANARD_VERSION_MAJOR                        1
+#define CANARD_VERSION_MINOR                        0
```
Comment (pavel-kirienko) [inline canard.h]: Are we sure we're ready to make this commitment? Just asking.

File: tests/demo.c [hunk 4]
```diff
@@ -27,16 +27,13 @@
 
 /*
  * Some useful constants defined by the UAVCAN specification.
- * Data type signature values can be easily obtained with the script show_data_type_info.py
  */
-#define UAVCAN_NODE_ID_ALLOCATION_DATA_TYPE_ID                      1
-#define UAVCAN_NODE_ID_ALLOCATION_DATA_TYPE_SIGNATURE               0x0b2a812620a11d40
+#define UAVCAN_NODE_ID_ALLOCATION_FIXED_SUBJECT_ID                  1
```
Comment (pavel-kirienko) [inline tests/demo.c]: None of the types used here are valid for UAVCAN v1.0. See:

- https://github.com/UAVCAN/dsdl/blob/uavcan-v1.0/uavcan/node/32085.Heartbeat.1.0.uavcan
- https://github.com/UAVCAN/dsdl/blob/uavcan-v1.0/uavcan/node/430.GetInfo.0.1.uavcan
- https://github.com/UAVCAN/dsdl/blob/uavcan-v1.0/uavcan/pnp/32742.NodeIDAllocationDataMTU8.0.1.uavcan (assuming CAN 2.0)

Should we update them now, or do you want to carry this change over into another PR?

File: dsdl_compiler/libcanard_dsdl_compiler/code_type_template.tmpl [hunk 5]
```diff
@@ -76,19 +76,19 @@ uint32_t ${type_name}_encode_internal(${type_name}* source,
                     %if f.bitlen < 8
     ${f.array_max_size_bit_len}
     //  - Add array length, last item, but bitlen < 8.
-    canardEncodeScalar(msg_buf, offset, ${f.array_max_size_bit_len}, (void*)&source->${'%s' % ((f.name + '.len'))});
+    canardEncodePrimitive(msg_buf, offset, ${f.array_max_size_bit_len}, (void*)&source->${'%s' % ((f.name + '.len'))});
     offset += ${f.array_max_size_bit_len};
                     %else
     if (! root_item)
```
Comment (pavel-kirienko) [inline dsdl_compiler/libcanard_dsdl_compiler/code_type_template.tmpl]: The TAO (tail array optimization) logic needs to be removed (finally!). Would you like to do it in a new PR?

File: canard.h [hunk 6]
```diff
@@ -164,15 +164,9 @@ typedef struct CanardTxQueueItem CanardTxQueueItem;
...
- * correct data type signature, otherwise transfer reception will fail with CRC mismatch error. Please refer to the
- * specification for more details about data type signatures. Signature for any data type can be obtained in many
- * ways; for example, using the command line tool distributed with Libcanard (see the repository).
  */
 typedef bool (* CanardShouldAcceptTransfer)(const CanardInstance* ins,          ///< Library instance
-                                            uint64_t* out_data_type_signature,  ///< Must be set by the application!
-                                            uint16_t data_type_id,              ///< Refer to the specification
+                                            uint16_t port_id,                   ///< Refer to the specification
                                             CanardTransferType transfer_type,   ///< Refer to CanardTransferType
                                             uint8_t source_node_id);            ///< Source node ID or Broadcast (0)
```
Comment (pavel-kirienko) [inline canard.h]:
```suggestion
                                            uint8_t source_node_id);            ///< Source node ID or Broadcast (255)
```

File: canard.h (hunk 3 above)
Comment (aasmune) [reply canard.h]: As I see it, until the uavcan-v1.0 is merged back to master it is still considered as a development branch and API may still be changed. I thought it was natural to update it here due to all the other renamings, but if you don't want to update the major version until we are 100% finished and ready for merge back to master that is fine for me.

File: tests/demo.c (hunk 4 above)
Comment (aasmune) [reply tests/demo.c]: I have created #107 to update the demo application to use v1.0. For now, I just updated the DATA_TYPE_ID to FIXED_SUBJECT_ID to conform with the other changes done in the PR.

File: canard.h (hunk 3 above)
Comment (pavel-kirienko) [reply canard.h]: Okay, it makes sense, let's keep it at 1.0.

File: dsdl_compiler/libcanard_dsdl_compiler/code_type_template.tmpl (hunk 5 above)
Comment (aasmune) [reply dsdl_compiler/libcanard_dsdl_compiler/code_type_template.tmpl]: Would prefer to do it in another PR, has created #108.

**Comment by aasmune** [comment]
> This is cool. This change alone is insufficient yet to make things compatible with UAVCAN v1.0, but I figured you want to split this work over multiple PRs. Please also note that there are minor code formatting issues where you have renamed functions.

Is it many changes which must be done to be compatible with UAVCAN v1.0 beside TAO and updating the demo application as mentioned? I must admit that I don't know all the changes needed for v0.1 -> v1.0 by heart atm.

**Comment by pavel-kirienko** [comment]
Off the top of my head, the only things to update are:

- TAO
- CRC
- Switch to new data types in the demo application

It would be also great to update the DSDL compiler to PyDSDL v0.4 (its API is almost stable now), but that is not really important right now and it doesn't affect wire compatibility.

File: canard.h (hunk 3 above)
Comment (pavel-kirienko) [reply canard.h]: While we're changing version numbers, may we also implement explicit differentiation between the version numbers of the library itself and of the specification that it implements? Like this: https://github.com/thirtytwobits/libuavcan/pull/1/files

Libcanard version 1.0 implements UAVCAN version 1.0, which is essentially a coincidence; at some point there may be libcanard v2.0 implementing UAVCAN v1.0, causing confusion among users unless we version it more explicitly. I don't think it's necessary to also specify the minor version of the standard (this decision can be reviewed later).

**Comment by aasmune** [comment]
As TAO and CRC will be handled in another PRs, I think we are ready for another review.

**Pull request merged**